engine/access/subscription/execution_data_tracker.go

```
@@ -0,0 +1,221 @@
+// if the start height is out of bounds based on indexed heights (when index is used).
+// - storage.ErrNotFound   - if a block is provided and does not exist.
+// - codes.Internal        - if there is an internal error.
+func (e *ExecutionDataTrackerImpl) GetStartHeight(ctx context.Context, startBlockID flow.Identifier, startHeight uint64) (uint64, error) {
+	var height uint64 = 0
+	var err error
+
+	if startBlockID != flow.ZeroID && startHeight > 0 {
+		return 0, status.Errorf(codes.InvalidArgument, "only one of start block ID and start height may be provided")
+	}
+
+	// get the start height based on the provided starting block ID
+	if startBlockID != flow.ZeroID {
+		height, err = e.GetStartHeightFromBlockID(startBlockID)
+	}
+
+	// get start height based on the provided starting block height
+	if startHeight > 0 {
+		height, err = e.GetStartHeightFromHeight(startHeight)
+	}
+
+	// get start height based latest sealed block when neither startBlockID nor startHeight is provided
+	if startBlockID == flow.ZeroID && startHeight == 0 {
+		height, err = e.GetStartHeightFromLatest(ctx)
+	}
+
+	if err != nil {
+		return 0, err
+	}
+
+	// ensure that the resolved start height is available
+	return e.checkStartHeight(height)
+}
+
+// GetHighestHeight returns the highest height that we have consecutive execution data for.
+func (e *ExecutionDataTrackerImpl) GetHighestHeight() uint64 {
+	return e.highestHeight.Value()
+}
+
+// OnExecutionData is used to notify the tracker when a new execution data is received.
+//
+// No errors expected during normal operations.
+func (e *ExecutionDataTrackerImpl) OnExecutionData(executionData *execution_data.BlockExecutionDataEntity) error {
+	header, err := e.headers.ByBlockID(executionData.BlockID)
+	if err != nil {
+		// if the execution data is available, the block must be locally finalized
+		return rpc.ConvertError(err, "failed to get header for block", codes.Internal)
```

if you remove this function from the engine, this should throw a fatal error